### PR TITLE
custom-levels - Add ability to load custom lands

### PR DESCRIPTION
**Pull request by raffclar**

### CODE REVIEW

File: src/3D/MeshPack.h [hunk 1]
```diff
@@ -64,6 +64,7 @@ class MeshPack
 	void loadTextures(const std::map<std::string, pack::G3DTexture>& textures);
 	void loadMeshes(const std::vector<std::vector<uint8_t>>& meshes);
 
+	bool _enableUnknownMeshes = false;
```
Comment (bwrsandman) [inline src/3D/MeshPack.h:67]: Could you move default values to the constructor to be more consistent?

File: src/Game.h [hunk 2]
```diff
@@ -40,6 +41,7 @@ class FileSystem;
 class GameWindow;
 class Gui;
 class EventManager;
+class LevelLocator;
```
Comment (bwrsandman) [inline src/Game.h:44]: Forward declare not necessary if you're including the header

File: src/LHScriptX/FeatureScriptCommands.cpp [hunk 3]
```diff
@@ -317,6 +318,12 @@ inline glm::vec3 GetSize(int size)
 	return glm::vec3(size, size, size) * 0.001f;
 }
 
+void FeatureScriptCommands::SetATownInfluenceMultiplier(int32_t townId, float mult)
+{
+	SPDLOG_LOGGER_ERROR(spdlog::get("scripting"), "LHScriptX: {}:{}: Function {} not implemented.", __FILE__, __LINE__,
```
Comment (bwrsandman) [inline src/LHScriptX/FeatureScriptCommands.cpp:323]: Could you check if `SPDLOG_LOGGER_ERROR` already includes the file and line?

File: src/Level.cpp [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#include "Level.h"
+
+using namespace openblack;
```
Comment (bwrsandman) [inline src/Level.cpp:3]: You can remove this file / doesn't include the copyright

File: src/LevelLocator.cpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#include "LevelLocator.h"
```
Comment (bwrsandman) [inline src/LevelLocator.cpp:1]: Missing copyright

File: src/LevelLocator.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#include <vector>
+
+#include "Level.h"
+
+namespace openblack
+{
+class LevelLocator
+{
+public:
+	void LoadLevels(FileSystem& fs);
```
Comment (bwrsandman) [inline src/LevelLocator.h:22]: can fs be const?

**Review by bwrsandman — CHANGES_REQUESTED** [review]

File: src/Level.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <string>
+#include <filesystem>
+
+namespace openblack
+{
+class Level
+{
+public:
+	Level(std::filesystem::path path) : _name(path.filename().string()), _scriptPath(path) {}
```
Comment (bwrsandman) [inline src/Level.h:20]: You could move path instead of taking a copy

File: src/Level.h [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	const std::string& GetName() const { return _name; };
```
Comment (bwrsandman) [inline src/Level.h:21]: This could be just `_scriptPath.filename().string()` instead of keeping a copy

**Review by bwrsandman — CHANGES_REQUESTED** [review]

File: src/LevelLocator.h (hunk 6 above)
Comment (raffclar) [reply src/LevelLocator.h]: Yep and the new fs function

File: src/Level.h (hunk 7 above)
Comment (raffclar) [reply src/Level.h]: So path is an r-value reference?

File: src/3D/MeshPack.h (hunk 1 above)
Comment (raffclar) [reply src/3D/MeshPack.h]: Done. Updated the constructor

File: src/LevelLocator.cpp (hunk 5 above)
Comment (raffclar) [reply src/LevelLocator.cpp:10]: Oops thanks

File: src/Level.h (hunk 7 above)
Comment (raffclar) [reply src/Level.h]: Using std::move from the caller

File: src/Level.h (hunk 8 above)
Comment (raffclar) [reply src/Level.h]: Good idea

**Comment by bwrsandman** [comment]
Is it normal that a bunch of these maps cause crashes? 
```
terminate called after throwing an instance of 'openblack::lhscriptx::LexerException'
  what():  unexpected character: �
```
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  File ./Data/Landscape/dance.led not found
```

Also, is there a way to sort these?
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  File ./Data/Landscape/dance.led not found
```

**Comment by raffclar** [comment]
@bwrsandman Here's what we could do 
* All .txt files in the playgrounds folder are treated as level files and are added to menubar dropdown.
* Campaign levels 1-5 and tutorial island are manually specified and added to the menubar dropdown.

There's no other way that I can see right now that would work.

**Comment by bwrsandman** [comment]
Is it possible to open the files to validate?

**Comment by raffclar** [comment]
We could create a `bool Script::IsScriptFile(fs::path& p)` function and use `bool Script::isCommand(const std::string& identifier)` over all line items.

File: src/Level.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Level.h:2]: 2022

File: src/LevelLocator.cpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/LevelLocator.cpp:2]: 2022

File: src/LevelLocator.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/LevelLocator.h:2]: 2022

File: src/Gui/Gui.cpp [hunk 12]
```diff
@@ -672,21 +673,20 @@ bool Gui::ShowMenu(Game& game)
...
-			constexpr std::array<std::pair<std::string_view, std::string_view>, 3> PlaygroundIslands = {
-			    std::pair {"Two Gods", "TwoGods.txt"},
-			    std::pair {"Three Gods", "ThreeGods.txt"},
-			    std::pair {"Four Gods", "FourGods.txt"},
-			};
+			auto& resources = Locator::resources::ref();
+			auto& levelsManager = resources.GetLevels();
+			auto campaigns = std::vector<entt::resource_handle<const Level>>();
+			auto playgrounds = std::vector<entt::resource_handle<const Level>>();
+			levelsManager.Each([&campaigns, &playgrounds](entt::id_type id, entt::resource_handle<const Level> level) {
```
Comment (bwrsandman) [inline src/Gui/Gui.cpp:680]: Warnings here

File: src/Level.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <filesystem>
+#include <string>
+
+namespace openblack
+{
+class Level
+{
+public:
+	Level(const std::string name, const std::filesystem::path path, bool isCampaign)
```
Comment (bwrsandman) [inline src/Level.h:20]: Copies don't need to be const.
You should also `std::move` them
Unless they are meant to be references

File: src/Game.cpp [hunk 14]
```diff
@@ -456,6 +457,42 @@ bool Game::Run()
...
+	levelManager.Load("campaign/Land1", "Land 1", scriptsPath / "Land1.txt", true);
+	levelManager.Load("campaign/Land2", "Land 2", scriptsPath / "Land2.txt", true);
+	levelManager.Load("campaign/Land3", "Land 3", scriptsPath / "Land3.txt", true);
+	levelManager.Load("campaign/Land4", "Land 4", scriptsPath / "Land4.txt", true);
+	levelManager.Load("campaign/Land5", "Land 5", scriptsPath / "Land5.txt", true);
+	// Load Playgrounds
+	levelManager.Load("playgrounds/LandT", "Tutorial Island", scriptsPath / "LandT.txt", false);
+	levelManager.Load("playgrounds/TwoGods", "Two Gods", scriptsPath / "Playgrounds" / "TwoGods.txt", false);
+	levelManager.Load("playgrounds/ThreeGods", "Three Gods", scriptsPath / "Playgrounds" / "ThreeGods.txt", false);
+	levelManager.Load("playgrounds/FourGods", "Four Gods", scriptsPath / "Playgrounds" / "FourGods.txt", false);
```
Comment (bwrsandman) [inline src/Game.cpp:471]: Can you put a todo for #400

File: src/Gui/Gui.cpp [hunk 15]
```diff
@@ -33,6 +33,7 @@
 #include <SDL2/SDL_syswm.h>
 #endif
 
+#include "Locator.h"
```
Comment (bwrsandman) [inline src/Gui/Gui.cpp:36]: Could you match the surrounding code and use `<>`

File: src/Gui/Gui.cpp [hunk 16]
```diff
@@ -672,21 +673,20 @@ bool Gui::ShowMenu(Game& game)
...
-			    std::pair {"Land 4", "Land4.txt"},
-			    std::pair {"Land 5", "Land5.txt"},
-			    std::pair {"God's Playground", "LandT.txt"},
-			    // clang-format on
-			};
-			constexpr std::array<std::pair<std::string_view, std::string_view>, 3> PlaygroundIslands = {
-			    std::pair {"Two Gods", "TwoGods.txt"},
-			    std::pair {"Three Gods", "ThreeGods.txt"},
-			    std::pair {"Four Gods", "FourGods.txt"},
-			};
```
Comment (bwrsandman) [inline src/Gui/Gui.cpp:689]: Nice!

File: src/Game.cpp [hunk 17]
```diff
@@ -456,6 +457,42 @@ bool Game::Run()
 		animationManager.Load(i, animations[i]);
 	}
 
+	// Load the campaign levels
+	const auto scriptsPath = _fileSystem->FindPath(_fileSystem->ScriptsPath());
+	levelManager.Load("campaign/Land1", "Land 1", scriptsPath / "Land1.txt", true);
+	levelManager.Load("campaign/Land2", "Land 2", scriptsPath / "Land2.txt", true);
+	levelManager.Load("campaign/Land3", "Land 3", scriptsPath / "Land3.txt", true);
+	levelManager.Load("campaign/Land4", "Land 4", scriptsPath / "Land4.txt", true);
+	levelManager.Load("campaign/Land5", "Land 5", scriptsPath / "Land5.txt", true);
```
Comment (bwrsandman) [inline src/Game.cpp:466]: If campaign don't have names, the names here are kind of random. 
Perhaps the name can be set at load time allowing us to add and remove campaign levels dynamically.

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Could you rebase on master. There might be new fails use to #387.
You might need to copy `test/mock/Scripts/Land1.txt` into Lands 1 to 5 and the playgrounds

File: src/Gui/Gui.cpp (hunk 12 above)
Comment (raffclar) [reply src/Gui/Gui.cpp]: Is this fixed? I'm not seeing it but I bet it's unused `id`

Comment (bwrsandman) [reply src/Gui/Gui.cpp]: You can see warnings in `Files Changed`

File: src/Game.cpp (hunk 17 above)
Comment (raffclar) [reply src/Game.cpp:467]: I could use the filename instead?

Comment (raffclar) [reply src/Game.cpp:467]: The problem is that `Scripts` folder contains many .txt files that are not levels. if we iterate over the Scripts directory looking for .txt files and we add them to the menu, they'll be a lot of noise.

**Comment by raffclar** [comment]
> Could you rebase on master. There might be new fails use to #387. You might need to copy `test/mock/Scripts/Land1.txt` into Lands 1 to 5 and the playgrounds

Yep :( It's failing because of this

File: src/Game.cpp (hunk 17 above)
Comment (bwrsandman) [reply src/Game.cpp:467]: All of the ones that aren't `InfoScript*.txt` are actually scene files though

File: src/Gui/Gui.cpp [hunk 18]
```diff
@@ -698,15 +698,15 @@ bool Gui::ShowMenu(Game& game)
 			auto& filesystem = Game::instance()->GetFileSystem();
```
Comment (bwrsandman) [inline src/Gui/Gui.cpp:698]: New warnings because filesystem is no longer used

**Review by bwrsandman — CHANGES_REQUESTED** [review]

**Review by bwrsandman — APPROVED** [review]

**Comment by bwrsandman** [comment]
The fail will go away if you rebase

**Pull request merged**